### ISSUE TITLE
Tensor::one_hot() drops autodiff from backend when original tensor is autodiff

### ISSUE THREAD

**Issue by fuzing** [tracel-ai/burn#5337]
I have a custom loss function used in training (currently using libtorch->CUDA backend).  When computing the  one hot encoding for the targets (which is initially a classification label index and lives on an Autodiff variant of the libtorch->CUDA backend), Autodiff is dropped from the computed tensor, causing subsequent operations with other "on-device" tensors to panic due to the incompatibility between autodiff and non-autodiff variants.

```
pub fn forward(&self, logits: Tensor<2>, targets: Tensor<1, Int>) -> Tensor<1> {
        let [_batch_size, num_classes] = logits.dims();

        // probabilities with clamping via epsilon at both the min and max
        let probs = softmax(logits.clone(), 1).clamp(self.eps, 1.0 - self.eps);

        // one-hot encode target indices -> [batch_size, num_classes]
        let targets_one_hot = targets.one_hot::<2>(num_classes);

        // reveals autodiff dropped from one hot tensor
        eprintln!("targets_one_hot: {}", targets_one_hot);

       // remainder of function removed for brevity
}
```

**Comment by fuzing** [comment]
Sorry - forgot to mention, I'm using the latest source version, as of this morning, directly from the repo.

**Comment by fuzing** [comment]
@laggui  - Amazing - thank you!

**Comment by laggui** [comment]
Thanks for flagging this issue by the way. Breaking away from the backend generics in the user API did remove that explicit contract (`B: Backend` vs `B: AutodiffBackend`), and it seems we might have missed this in the dispatch.

Linking #5131, and my comment https://github.com/tracel-ai/burn/pull/5340#pullrequestreview-4908997008 for reference.
